### PR TITLE
raft: better doc

### PR DESCRIPTION
**Pull request by xiang90**
@heyitsanthony @philips @gyuho

### CODE REVIEW

File: raft/README.md [hunk 1]
```diff
@@ -5,6 +5,14 @@ The state machine is kept in sync through the use of a replicated log.
 For more details on Raft, see "In Search of an Understandable Consensus Algorithm"
 (https://ramcloud.stanford.edu/raft.pdf) by Diego Ongaro and John Ousterhout.
 
+This Raft library is stable and feature completed. It is **the most widely used** (as of 2016) Raft library that serves tens of thousands clusters each day. It powers distributed systems like etcd, Kubernetes, Docker Swarm, Cloud Foundry Diego, CockroachDB, TiDB, Project Calico, Flannel and more.
+
+Most Raft implementations has a monolithic design, which includes storage, messaging serialization, transport. This Raft library takes a very different approach by only implementing the core algorithm part of raft. The design philosophy of this raft library is minimalism, flexibility, determinism, and performance.
```
Comment (gyuho) [inline raft/README.md]: s/`has`/`have`/?

File: raft/README.md [hunk 2]
```diff
@@ -5,6 +5,14 @@ The state machine is kept in sync through the use of a replicated log.
 For more details on Raft, see "In Search of an Understandable Consensus Algorithm"
 (https://ramcloud.stanford.edu/raft.pdf) by Diego Ongaro and John Ousterhout.
 
+This Raft library is stable and feature completed. It is **the most widely used** (as of 2016) Raft library that serves tens of thousands clusters each day. It powers distributed systems like etcd, Kubernetes, Docker Swarm, Cloud Foundry Diego, CockroachDB, TiDB, Project Calico, Flannel and more.
```
Comment (heyitsanthony) [inline raft/README.md]: This Raft library is stable and feature complete. As of 2016, it is **the most widely used** Raft library in production, serving tens of thousands clusters each day. It powers distributed systems such as etcd, Kubernetes, Docker Swarm, Cloud Foundry Diego, CockroachDB, TiDB, Project Calico, Flannel, and more.

File: raft/README.md (hunk 1 above)
Comment (heyitsanthony) [inline raft/README.md]: Most Raft implementations have a monolithic design, including storage handling, messaging serialization, and network transport. This library instead follows a minimalistic design philosophy by only implementing the core raft algorithm. This minimalism buys flexibility, determinism, and performance.

File: raft/README.md [hunk 3]
```diff
@@ -5,6 +5,14 @@ The state machine is kept in sync through the use of a replicated log.
...
+
+Most Raft implementations has a monolithic design, which includes storage, messaging serialization, transport. This Raft library takes a very different approach by only implementing the core algorithm part of raft. The design philosophy of this raft library is minimalism, flexibility, determinism, and performance.
+
+To minimize the library as well as provide flexibility, it only implements the Raft algorithm part and leave the network IO and disk IO parts to its users. Users need to implement their own transportation layer to enable actual message passing between Raft peers over wire. Also users need to implement their own storage layer to persist Raft logs and states.
```
Comment (heyitsanthony) [inline raft/README.md]: To keep the codebase small as well as provide flexibility, the library only implements the Raft algorithm; both network and disk IO are left to the user. Library users must implement their own transportation layer for message passing between Raft peers over the wire. Similarly, users must implement their own storage layer to persist the Raft log and state.

File: raft/README.md [hunk 4]
```diff
@@ -5,6 +5,14 @@ The state machine is kept in sync through the use of a replicated log.
...
+
+To make the Raft library easy to be tested, It is important to make the behavior of Raft deterministic. We model Raft as a state machine. The input of the state machine is a `Message`. A message can be a normal network message from remote peers or a timing related message. The output of the state machine is an array of `Messages`, `log entries` and `Raft state changes`. For the state machine with same state `A`, same input `M` should always generate same output `{[]Messages, []LogEntries, NextState}`.
```
Comment (heyitsanthony) [inline raft/README.md]: In order to easily test the Raft library, its behavior should be deterministic. To achieve this determinism, the library models Raft as a state machine.  The state machine takes a `Message` as input. A message can either be a local timer update or a network message sent from a remote peer. The state machine's output is a 3-tuple, `{[]Messages, []LogEntries, NextState}` consisting of an array of `Messages`, `log entries`, and `Raft state changes`. For state machines with the same state `A`, the same state machine input should always generate the same state machine output.

**Comment by xiang90** [comment]
@heyitsanthony All fixed. PTAL.

File: raft/README.md [hunk 5]
```diff
@@ -5,6 +5,14 @@ The state machine is kept in sync through the use of a replicated log.
...
+
+To keep the codebase small as well as provide flexibility, the library only implements the Raft algorithm; both network and disk IO are left to the user. Library users must implement their own transportation layer for message passing between Raft peers over the wire. Similarly, users must implement their own storage layer to persist the Raft log and state.
+
+In order to easily test the Raft library, its behavior should be deterministic. To achieve this determinism, the library models Raft as a state machine.  The state machine takes a `Message` as input. A message can either be a local timer update or a network message sent from a remote peer. The state machine's output is a 3-tuple, `{[]Messages, []LogEntries, NextState}` consisting of an array of `Messages`, `log entries`, and `Raft state changes`. For state machines with the same state `A`, the same state machine input should always generate the same state machine output.
```
Comment (heyitsanthony) [inline raft/README.md]: s/3-tuple,/3-tuple/
s/with the same machine state A,/with the same machine state,/

**Comment by heyitsanthony** [comment]
lgtm after fixing final nits

**Pull request merged**